lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,551 @@
+	} while (is_index_ahead(parser, index) && !err);
+
+	return err;
+}
+
+int at_parser_init(struct at_parser *parser, const char *at)
+{
+	if (!parser || !at) {
+		return -EINVAL;
+	}
+
+	memset(parser, 0, sizeof(struct at_parser));
+
+	parser->at = at;
+	parser->cursor = at;
+	parser->init_sentinel = INIT_SENTINEL;
+
+	return 0;
+}
+
+int at_parser_next(struct at_parser *parser)
+{
+	int err;
+	struct at_token token = {0};
+
+	err = at_parser_check(parser);
+	if (err) {
+		return err;
+	}
+
+	do {
+		err = at_parser_tok(parser, &token);
+	} while (!err);
+
+	if (err != -EAGAIN) {
+		return -EOPNOTSUPP;
+	}
+
+	trim_crlf(&parser->cursor);
+
+	/* Reset count. */
+	parser->count = 0;
+	/* Set pointer of current AT command string to the current cursor, which points to the
+	 * beginning of a new AT command line.
+	 */
+	parser->at = parser->cursor;
+
+	return 0;
+}
+
+int at_parser_cmd_get(struct at_parser *parser, const char **at)
+{
+	int err;
+
+	if (!at) {
+		return -EINVAL;
+	}
+
+	err = at_parser_check(parser);
+	if (err) {
+		return err;
+	}
+
+	*at = parser->at;
+
+	return 0;
+}
+
+int at_parser_cmd_type_get(struct at_parser *parser, enum at_parser_cmd_type *type)
+{
+	int err;
+	struct at_token tok;
+
+	if (!type) {
+		return -EINVAL;
+	}
+
+	err = at_parser_check(parser);
+	if (err) {
+		return err;
+	}
+
+	tok = at_match_cmd(parser->at, NULL);
+	switch (tok.type) {
+	case AT_TOKEN_TYPE_CMD_TEST:
+		*type = AT_PARSER_CMD_TYPE_TEST;
+		break;
+	case AT_TOKEN_TYPE_CMD_READ:
+		*type = AT_PARSER_CMD_TYPE_READ;
+		break;
+	case AT_TOKEN_TYPE_CMD_SET:
+		*type = AT_PARSER_CMD_TYPE_SET;
+		break;
+	default:
+		*type = AT_PARSER_CMD_TYPE_UNKNOWN;
+		return -EOPNOTSUPP;
+	}
+
+	return 0;
+}
+
+int at_parser_cmd_count_get(struct at_parser *parser, size_t *count)
+{
+	int err;
+	struct at_token token = {0};
+
+	if (!count) {
+		return -EINVAL;
+	}
+
+	err = at_parser_check(parser);
+	if (err) {
+		return err;
+	}
+
+	do {
+		err = at_parser_tok(parser, &token);
+	} while (!err);
+
+	*count = parser->count;
+
+	return (err == -EIO || err == -EAGAIN) ? 0 : err;
+}
+
+static int at_parser_num_get_impl(struct at_parser *parser, size_t index, void *value,
+				  enum at_num_type type)
+{
+	int err;
+	struct at_token token = {0};
+
+	if (!value) {
+		return -EINVAL;
+	}
+
+	err = at_parser_check(parser);
+	if (err) {
+		return err;
+	}
+
+	err = at_parser_seek(parser, index, &token);
+	if (err) {
+		return err;
+	}
+
+	if (token.type != AT_TOKEN_TYPE_INT) {
+		return -EOPNOTSUPP;
+	}
+
+	errno = 0;
+	char *next = NULL;
+
+	/* Check unsigned 64-bit integer first, using its own parsing function. */
+	if (type == AT_NUM_TYPE_UINT64) {
+		if (token.start[0] == MINUS_SIGN) {
+			return -ERANGE;
+		}
+
+		uint64_t val = strtoull(token.start, &next, 10);
+
+		if (errno == ERANGE) {
+			return -ERANGE;
+		}
+
+		*(uint64_t *)value = val;
+
+		return 0;
+	}
+
+	int64_t val = strtoll(token.start, &next, 10);
+
+	if (errno == ERANGE) {
+		return -ERANGE;
+	}
```

Comment:
Why are we doing this instead of handling this in the `switch` case below?

Reply:
Updated.